### PR TITLE
chore(gatsby): Optimise create page action validation

### PR DESCRIPTION
**Pull request by ascorbic**
This PR optimises the validation of createPage calls. It extracts the validation into a helper function, and memoizes most of the actions.

This is part of the develop state machine refactor

### CODE REVIEW

File: packages/gatsby/src/utils/validate-page-component.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  }
+  if (validationCache.has(component)) {
+    return {}
+  }
+  if (!path.isAbsolute(component)) {
+    return {
+      error: {
+        id: `11326`,
+        context: {
+          pluginName: name,
```
Comment (pieh) [inline packages/gatsby/src/utils/validate-page-component.ts:29]: We are not passing `name` here from `createPage` action

Comment (ascorbic) [reply packages/gatsby/src/utils/validate-page-component.ts:29]: Weird that TypeScript didn't catch this. It seems to have randomly thought it was an ambient type from the dom lib

File: packages/gatsby/src/utils/validate-page-component.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
+import path from "path"
+import fs from "fs-extra"
+import { IGatsbyPage } from "../redux/types"
+
+const validationCache = new Map<string, boolean>()
```
Comment (wardpeet) [inline packages/gatsby/src/utils/validate-page-component.ts:5]: A set seems more appropriate.

```suggestion
const validationCache = new Set<string>()
```

File: packages/gatsby/src/utils/validate-page-component.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+          context: {
+            component,
+          },
+        },
+        panicOnBuild: true,
+      }
+    }
+  }
+
+  validationCache.set(component, true)
```
Comment (wardpeet) [inline packages/gatsby/src/utils/validate-page-component.ts:99]:
```suggestion
  validationCache.add(component)
```

**Pull request merged**